Cash Flow Report

ccfiel opened the pull request:
@anandpdoshi Cash Flow is done! :)

ccfiel committed:
Cash Flow Report

ccfiel committed:
Cash Flow Report

ccfiel committed:
Cash Flow Report

anandpdoshi commented on erpnext/accounts/doctype/account/account.json:
```diff
@@ -261,9 +271,10 @@
    "oldfieldname": "account_type", 
    "oldfieldtype": "Select", 
-   "options": "\nBank\nCash\nTax\nChargeable\nWarehouse\nReceivable\nPayable\nEquity\nFixed Asset\nCost of Goods Sold\nExpense Account\nRound Off\nIncome Account\nStock Received But Not Billed\nExpenses Included In Valuation\nStock Adjustment\nStock\nTemporary", 
+   "options": "\nBank\nCash\nDepreciation\nTax\nChargeable\nWarehouse\nReceivable\nPayable\nEquity\nFixed Asset\nCost of Goods Sold\nExpense Account\nRound Off\nIncome Account\nStock Received But Not Billed\nExpenses Included In Valuation\nStock Adjustment\nStock\nTemporary", 
```
@nabinhait any concerns about this?

anandpdoshi commented on erpnext/accounts/report/financial_statements.py:
```diff
@@ -258,3 +258,35 @@ def get_columns(period_list):
+        gl_sum = frappe.db.sql_list("""select sum(credit) - sum(debit) FROM `tabGL Entry`
+                                    where company='%s' and posting_date >= '%s' and posting_date <= '%s' and
+                                    account in ( SELECT name FROM tabAccount WHERE account_type = '%s')""" %
+                                    (company, period['from_date'], period['to_date'], account_type))
```
possible injection

ccfiel replied:
@anandpdoshi  oh i see do i need to change this? or can you do it in your side :)

anandpdoshi replied:
You can do it if you find some time. Else Nabin will look into it on Monday. 

Sent from my phone

> On 12-Dec-2015, at 10:01 AM, Chris Ian Fiel <notifications@github.com> wrote:
> 
> In erpnext/accounts/report/financial_statements.py:
> 
> > @@ -258,3 +258,35 @@ def get_columns(period_list):
> >  		})
> >  
> >  	return columns
> > +
> > +
> > +def get_data_account_type(company, account_type, period_list):
> > +    data = {}
> > +    for period in period_list:
> > +        gl_sum = frappe.db.sql_list("""select sum(credit) - sum(debit) FROM `tabGL Entry`
> > +                                    where company='%s' and posting_date >= '%s' and posting_date <= '%s' and
> > +                                    account in ( SELECT name FROM tabAccount WHERE account_type = '%s')""" %
> > +                                    (company, period['from_date'], period['to_date'], account_type))
> @anandpdoshi oh i see do i need to change this? or can you do it in your side :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

ccfiel replied:
@anandpdoshi I tried to look other sql_list code in erpnext. This is at accounts/utils.py line 252.
```
def remove_against_link_from_jv(ref_type, ref_no):
	linked_jv = frappe.db.sql_list("""select parent from `tabJournal Entry Account`
		where reference_type=%s and reference_name=%s and docstatus < 2""", (ref_type, ref_no))
```

We have the same syntax in making an sql call. Can you give me some idea how to avoid possible injection?

anandpdoshi replied:
Mapping values using `%` instead of passing values to the sql function can lead to injection, as string escaping isn't performed in the former case.

nabinhait replied:
I have fixed this in my local.

anandpdoshi commented:
@ccfiel Great work! Nabin might make some changes to make it generic.

@nabinhait Can you review this? Also, you will need to checkout the code via command line and make minor changes. Also note that the sections are hardcoded.

ccfiel commented:
@anandpdoshi @nabinhait  section/segment is hardcoded because cash flow has only 3 activities operating, investing and financing https://en.wikipedia.org/wiki/Cash_flow_statement. I think we just need to add more account type ( if I forgot any to include) under this 3 sections/segment that would affect the cash flow that is why I made an array for each section account type so that i can easily be added.

ccfiel commented:
@nabinhait any comments? 

cc: @anandpdoshi

inventobd commented:
Seriously ? 

Thanks a lot @ccfiel . A much needed feature.

nabinhait commented:
@ccfiel Awesome work!
Let me learn the Cash Flow Statement and test it in my local.

revant commented:
I didn't see "Net Change in Fixed Asset" because account_type of "Fixed Asset Warehouse" account was "Warehouse"

nabinhait commented:
@ccfiel I have some doubt.

1. Depreciation account is an expense account, it is debited while booking depreciation. Thats why Depreciation amount is coming as negative, I think it should be positive and added to the "Net Profit and Loss".

2. Related to depreciation again. While booking depreciation "Fixed Asset" accounts are credited. Thats why while showing "Net change in Fixed Asset" is also considering those credit values. I think it should only show the amount of only "Purchase amount for new fixed asset".

3. Should I add another account type "Write Off"? Write off amount should be reduced from "Net Profit / Loss".

Note: I am just a beginner on this topic.

nabinhait commented:
@revant You should change the account type for Fixed Asset Warehouse. You can select parent account manually while creating new warehouse.

revant commented:
If I change Account Type from "Warehouse" to "Fixed Asset" for Fixed Asset Warehouse Account I get error on submitting purchase receipt
```
No accounting entries for the following warehouses: 
Fixed Asset Warehouse - MNT
```
and there is no way to change back account_type from Fixed Asset to Warehouse to correct the purchase receipt again.

nabinhait commented:
@revant Actually its correct, fixed assets does not follow perpetual inventory system, hence no accounting entry needed in Purchase Receipt. In Purchase Invoice, Fixed Asset Account should be selected as "Expense Account", to book fixed asset value. You can set default expense account in item master as fixed asset account.

ccfiel commented:
@nabinhait  
1. yes you are correct I will issue a PR for this.
2. My suggestion is create a depreciation asset and leave the account type blank but you can still place it under fixed asset group in this way it will not be included in the computation in cash flow but it will be in the balance sheet
3. yes correct

ccfiel committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into develop

ccfiel committed:
By: nabinhait

Depreciation account is an expense account, it is debited while booking depreciation. Thats why Depreciation amount is coming as negative, I think it should be positive and added to the "Net Profit and Loss".

ccfiel commented:
@nabinhait I have already change the depreciation. You please check if does what you mean. Thanks

nabinhait commented:
@ccfiel Thanks. Replaced spaces with tabs and fixed injection issue via https://github.com/frappe/erpnext/commit/8b33a6cdc0e7b901573865f9b8bb982207b9ae45
